dvc/repo/experiments/__init__.py

```diff
@@ -786,6 +787,51 @@ def on_diverged(_ref: str, _checkpoint: bool):
                 pass
         return result
 
+    def get_revs_from_args(
+        self,
+        head_revs: Optional[Iterable[str]] = None,
+        num: int = 1,
+        all_branches: bool = False,
+        all_tags: bool = False,
+        all_commits: bool = False,
+    ) -> Mapping[str, None]:
+        from dvc.exceptions import InvalidArgumentError
+        from dvc.scm import RevError, resolve_rev
+
+        if num < 1 and num != -1:
+            raise InvalidArgumentError(f"Invalid number of commits '{num}'")
+
+        if not any([head_revs, all_branches, all_tags, all_commits]):
+            return {}
+
+        head_revs = head_revs or []
+        revs = []
+        for rev in head_revs:
+            n = 0
+            while True:
+                if num == n:
+                    break
+                try:
+                    head = fix_exp_head(self.repo.scm, f"{rev}~{n}")
+                    assert head
+                    revs.append(resolve_rev(self.repo.scm, head))
+                except RevError:
+                    break
+                n += 1
+
+        return_revs = OrderedDict(
+            (rev, None)
+            for rev in self.repo.brancher(
```

Comment:
Using `brancher` here is a pretty heavy dependency since it creates the git-fs and changes the current `repo` instance for each rev. It's also bit redundant since we also use `brancher` on single revisions in `_collect_experiment_commit` (where we actually do want the git-fs and modified `repo` instance).

Since at this point in the call we only want the list of revisions that the combination of `all_branches/all_tags/all_commits` would give, what we probably want is a to split the current `brancher` functionality for just getting the revisions into a helper method https://github.com/iterative/dvc/blob/ceeda6d3072fb7d862518a60e9e2f6a0ae2cf3d2/dvc/repo/brancher.py#L30-L59

(so that we can get the revs but skip the git-fs/repo modification when we don't need it)

Reply:
Something like `iter_revs`  perhaps?

```python
for rev in iter_revs(revs, all_branches=all_branches, all_tags=all_tags, all_experiments=all_experiments, all_commits=all_commits):
    pass
```

Reply:
It is OK to rename `get_revs_from_flags` as `iter_revs` for me, but the problem preventing me merge the logic between `get_revs_from_flags` and `brancher` here is that there are some problems with the output format of `iter_revs`. 
    The output format of `brancher` without a `sha_only` flag can be `shas`, `branch names`, `tag names` or even `HEAD` or `HEAD~~~`. Besides this when one `commit` have multiple names among this, we need to the one that users provide to make them more understandable to the users. 
   While on the `get_revs_from_flags` side the output is quite simple, we just crunch all different inputs into `shas`.
   I think it is better to follow the output rule in `brancher` here for both of the two functions.

Reply:
The iterator can return both the `sha` and `names` pairs. It's really a `brancher()` design flaw that it's only designed to return either SHAs or names, but not both. see updated comment: https://github.com/iterative/dvc/pull/7204#discussion_r778556617